sdk/schemaregistry/schema-registry-avro/src/schemaRegistryAvroSerializer.ts

```diff
@@ -0,0 +1,226 @@
+  constructor(
+    registry: SchemaRegistry,
+    schemaGroup: string,
+    options?: SchemaRegistryAvroSerializerOptions
+  ) {
+    this.registry = registry;
+    this.schemaGroup = schemaGroup;
+    this.autoRegisterSchemas = options?.autoRegisterSchemas ?? false;
+  }
+
+  /** The schema group to use when making requests to the registry*/
+  public readonly schemaGroup: string;
+
+  /**
+   * Schema Registry where schemas are registered and obtained.
+   * Usually this is a SchemaRegistryClient instance.
+   */
+  public readonly registry: SchemaRegistry;
+
+  /**
+   * When true, register new schemas passed to serialize. Otherwise, and by
+   * default, fail if schema has not already been registered.
+   *
+   * Automatic schema registration is NOT recommended for production scenarios.
+   */
+  public autoRegisterSchemas: boolean;
+
+  // REVIEW: signature.
+  //
+  // - Better to serialize into a stream? I aborted that for now as I wanted to
+  //   do the simplest thing that could possibly work first to make sure there
+  //   were no blockers in our dependencies. I also wanted to get feedback on
+  //   what the API shape should be before diving into that.
+  //
+  // - This type should ultimately be able to implement a core ObjectSerializer
+  //   interface. Do we know what that would look like? Maybe it takes `any` as
+  //   the format-specific schema/type info? Or does it always take a
+  //   format-specific schema string?
+  //
+  //   The C#/Java approach of passing Type and assuming every serializer can
+  //   get its schema by reflecting on the type does not work for JavaScript. We
+  //   need to support arbitrary objects that match a schema.
+  //
+  //   Maybe each format expects a different property on this arg so that you
+  //   could at least pass enough info for multiple formats, and then your call
+  //   to ObjectSerializer is at least not tied to a single format?
+  //
+  // - Should we wrap all errors thrown by avsc to avoid having our exception //
+  //   contract being tied to its implementation details?
+  /**
+   * Serializes a value into a buffer.
+   *
+   * @param value The value to serialize.
+   * @param schema The Avro schema to use.
+   * @returns A new buffer with the serialized value
+   */
+  async serialize(value: any, schema: string): Promise<Buffer> {
+    const entry = await this.getSchemaByContent(schema);
+    const payload = entry.type.toBuffer(value);
+    const buffer = Buffer.alloc(PAYLOAD_OFFSET + payload.length);
+
+    buffer.writeUInt32BE(FORMAT_INDICATOR, 0);
+    buffer.write(entry.id, SCHEMA_ID_OFFSET, SCHEMA_ID_LENGTH, "utf-8");
+    payload.copy(buffer, PAYLOAD_OFFSET);
+    return buffer;
+  }
+
+  // REVIEW: signature. See serialize and s/serialize into/deserialize from/.
+  /**
+   * Deserializes a value from a buffer.
+   *
+   * @param buffer The buffer with the serialized value.
+   * @return The deserialized value.
+   */
+  async deserialize(buffer: Buffer): Promise<any> {
```

Comment:
```suggestion
  async deserialize<T>(buffer: Buffer): Promise<T> {
```

Reply:
Interesting. I didn't realize you could call a method defined that way without a type argument and nothing in the parameters to infer it from. (C# bias.) I see you get `unknown` back in that case.

My concern with this change, which may be off due to my lack of familiarity with TypeScript idioms, is that it moves a cast into our code, and might give an illusion of safety that isn't there?